sycl/include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp

```diff
@@ -0,0 +1,301 @@
+#define __OPENCL_CONSTANT_AS__ __attribute__((opencl_constant))
+#define __OPENCL_PRIVATE_AS__ __attribute__((opencl_private))
+#else
+#define __OPENCL_GLOBAL_AS__
+#define __OPENCL_GLOBAL_DEVICE_AS__
+#define __OPENCL_GLOBAL_HOST_AS__
+#define __OPENCL_LOCAL_AS__
+#define __OPENCL_CONSTANT_AS__
+#define __OPENCL_PRIVATE_AS__
+#endif
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+namespace detail {
+
+// Type-trait for checking if a type defines `operator[]`.
+template <typename T, typename = void>
+struct HasSubscriptOperator : std::false_type {};
+
+template <typename T>
+struct HasSubscriptOperator<
+    T, typename std::enable_if_t<
+           !std::is_void<decltype(std::declval<T>().operator[](0))>::value>>
+    : std::true_type{};
+
+} // namespace detail
+
+// Deduction guide
+template <typename T, typename... Args>
+annotated_arg(T, Args... args)
+    -> annotated_arg<T, detail::properties_t<Args...>>;
+
+template <typename T, typename... Args>
+annotated_arg(T, properties<std::tuple<Args...>>)
+    -> annotated_arg<T, detail::properties_t<Args...>>;
+
+template <typename T, typename old, typename... ArgT>
+annotated_arg(annotated_arg<T, old>, properties<std::tuple<ArgT...>>)
+    -> annotated_arg<
+        T, detail::merged_properties_t<old, detail::properties_t<ArgT...>>>;
+
+template <typename T, typename PropertyListT = detail::empty_properties_t>
+class annotated_arg {
+  // This should always fail when instantiating the unspecialized version.
+  static_assert(is_property_list<PropertyListT>::value,
+                "Property list is invalid.");
+};
+
+// Partial specialization for pointer type
+template <typename T, typename... Props>
+class __SYCL_SPECIAL_CLASS
+__SYCL_TYPE(annotated_arg) annotated_arg<T *, detail::properties_t<Props...>> {
+  using property_list_t = detail::properties_t<Props...>;
+  __OPENCL_GLOBAL_AS__ T *obj;
+
+  template <typename T2, typename PropertyListT> friend class annotated_arg;
+
+#ifdef __SYCL_DEVICE_ONLY__
+  void __init([[__sycl_detail__::add_ir_attributes_kernel_parameter(
+      detail::PropertyMetaInfo<Props>::name...,
+      detail::PropertyMetaInfo<Props>::value...)]] __OPENCL_GLOBAL_AS__ T
+                  *_obj) {
+    obj = _obj;
+  }
+#endif
+
+public:
+  static_assert(std::is_trivially_destructible<T>::value,
+                "Type T must be trivially destructible.");
+  static_assert(is_property_list<property_list_t>::value,
+                "Property list is invalid.");
+  static_assert(detail::SortedAllUnique<std::tuple<Props...>>::value,
+                "Duplicate properties in property list.");
+
+  annotated_arg() noexcept = default;
+  annotated_arg(const annotated_arg &) = default;
+  annotated_arg &operator=(annotated_arg &) = default;
+
+  annotated_arg(T *_ptr,
+                const property_list_t &PropList = properties{}) noexcept
+      : obj((__OPENCL_GLOBAL_AS__ T *)_ptr) {}
```

Comment:
A cast like this could cause warnings. Instead you can use `sycl::detail::cast_AS` which should make the right conversions for you.